protocol/app/prepare/full_node_prepare_proposal_test.go

```diff
@@ -1,220 +1,67 @@
-			clobEncoder: passingTxEncoderFour,
-		},
-		"Error: AddOtherTxs (additional) return error": {
-			maxBytes: 15,
-			txs:      [][]byte{{9, 8}, {9}, {}, {}},
-
-			pricesResp:    &pricestypes.MsgUpdateMarketPrices{},
-			pricesEncoder: passingTxEncoderFour,
-
-			fundingResp:    &perpetualtypes.MsgAddPremiumVotes{},
-			fundingEncoder: passingTxEncoderFour,
-
-			clobResp:    &clobtypes.MsgProposedOperations{},
-			clobEncoder: passingTxEncoderFour,
-		},
-		"Valid: Not all Others than can fit": {
-			maxBytes: 13,
-			txs:      [][]byte{{9}, {9}, {9}},
-
-			pricesResp:    &pricestypes.MsgUpdateMarketPrices{},
-			pricesEncoder: passingTxEncoderFour,
-
-			fundingResp:    &perpetualtypes.MsgAddPremiumVotes{},
-			fundingEncoder: passingTxEncoderFour,
-
-			clobResp:    &clobtypes.MsgProposedOperations{},
-			clobEncoder: passingTxEncoderFour,
-		},
-		"Valid: Additional Others fit": {
-			maxBytes: 15,
-			txs:      [][]byte{{9}, {9, 8}, {9, 8, 7}},
-
-			pricesResp:    &pricestypes.MsgUpdateMarketPrices{},
-			pricesEncoder: passingTxEncoderFour,
-
-			fundingResp:    &perpetualtypes.MsgAddPremiumVotes{},
-			fundingEncoder: passingTxEncoderFour,
-
-			clobResp:    &clobtypes.MsgProposedOperations{},
-			clobEncoder: passingTxEncoderFour,
-		},
+	var logBuffer bytes.Buffer
+	appOpts := map[string]interface{}{
+		flags.NonValidatingFullNodeFlag: true,
+		testApp.LoggerInstanceForTest:   log.TestingLoggerWithOutput(&logBuffer),
 	}
-
-	for name, tc := range tests {
-		t.Run(name, func(t *testing.T) {
-			mockPricesKeeper := mocks.PreparePricesKeeper{}
-			mockPricesKeeper.On("GetValidMarketPriceUpdates", mock.Anything).
-				Return(tc.pricesResp)
-
-			mockPerpKeeper := mocks.PreparePerpetualsKeeper{}
-			mockPerpKeeper.On("GetAddPremiumVotes", mock.Anything, mock.Anything).
-				Return(tc.fundingResp)
-
-			mockClobKeeper := mocks.PrepareClobKeeper{}
-			mockClobKeeper.On("GetOperations", mock.Anything, mock.Anything).
-				Return(tc.clobResp)
-
-			ctx, _, _, _, _, _ := keepertest.PricesKeepers(t)
-
-			handler := prepare.FullNodePrepareProposalHandler()
-
-			req := abci.RequestPrepareProposal{
-				Txs:        tc.txs,
-				MaxTxBytes: tc.maxBytes,
-			}
-
-			response := handler(ctx, req)
-			require.Equal(t, [][]byte{}, response.Txs)
-		})
+	tApp := testApp.NewTestAppBuilder().WithTesting(t).WithAppCreatorFn(testApp.DefaultTestAppCreatorFn(appOpts)).Build()
+
+	found := false
+	tApp.AdvanceToBlock(2, testApp.AdvanceToBlockOptions{
+		BlockTime:                         time.Time{},
+		RequestPrepareProposalTxsOverride: [][]byte{{9}, {9, 8}, {9, 8, 7}},
+		ValidateRespPrepare: func(context sdktypes.Context, proposal abci.ResponsePrepareProposal) (haltChain bool) {
+			require.Empty(t, proposal.Txs)
+			return true
+		},
+	})
+
+	for _, metrics := range sink.Data() {
+		metrics.RLock()
+		defer metrics.RUnlock()
+
+		if metric, ok := metrics.Counters["testService.prepare_proposal.handler.error.count;detail=prepare_proposal_txs"]; ok {
```

Comment:
noob question: when do we expect `ok` to be false? is this iterating all the metrics that's been generated?

Reply:
Yes, it is iterating all the metrics that have been produced and there are a few through this.